### PR TITLE
8290464: Optimize ResourceArea zapping on ResourceMark release

### PR DESCRIPTION
**Pull request by shipilev**
We have `ResourceMark` scopes where we either do not resource-allocate at all (see the bug for the example), or we do not allocate a lot in the arena. Still, in debug builds, we are zapping the entirety of `ResourceArea` chunk in those cases. This wastes testing cycles unnecessarily. Doing this in a bit smarter way -- zapping only the parts that were actually allocated in the chunk -- makes testing significantly faster.

Linux x86_64 fastdebug, `tier1`:

```
# Before
real 26m33.621s
user 1156m56.676s
sys 52m18.291s

# After
real 25m29.573s  ; -4.0%
user 1089m27.179s ; -5.8%
sys 52m17.630s
```

Additional testing:
 - [x] Linux x86_64 fastdebug, `FieldSetAccessibleTest` (~12% faster)
 - [x] Linux x86_64 release, `FieldSetAccessibleTest` (no regressions)
 - [x] Linux x86_64 fastdebug, `hotspot:tier1` (~6% faster)
 - [x] Linux x86_64 release, `hotspot:tier1` (no regressions)
 - [x] Linux x86_64 fastdebug, `tier1` (~4% faster)
 - [x] Linux x86_64 release, `tier1` (no regressions)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8290464](https://bugs.openjdk.org/browse/JDK-8290464): Optimize ResourceArea zapping on ResourceMark release

### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)
 * [Zhengyu Gu](https://openjdk.org/census#zgu) (@zhengyu123 - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9543/head:pull/9543` \
`$ git checkout pull/9543`

Update a local copy of the PR: \
`$ git checkout pull/9543` \
`$ git pull https://git.openjdk.org/jdk pull/9543/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9543`

View PR using the GUI difftool: \
`$ git pr show -t 9543`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9543.diff">https://git.openjdk.org/jdk/pull/9543.diff</a>

</details>

### CODE REVIEW

**Comment by shipilev** [comment]
> I think the placement of ResourceMark should be moved as well (maybe in a separate PR). 

All right, that would be #9548.

I actually thought about introducing `assert_with_resources` too, but reckoned if we want to have a global change, we better off optimizing the `ResourceMark` itself. Otherwise we would have to remember to use `assert_with_resource` every time...

File: src/hotspot/share/memory/resourceArea.hpp [hunk 1]
```diff
@@ -120,13 +122,28 @@ class ResourceArea: public Arena {
...
       assert(size_in_bytes() == state._size_in_bytes, "Sanity check");
     }
-    _chunk = state._chunk;      // Roll back to saved chunk.
-    _hwm = state._hwm;
-    _max = state._max;
 
-    // Clear out this chunk (to detect allocation bugs)
-    if (ZapResourceArea) {
-      memset(state._hwm, badResourceValue, state._max - state._hwm);
+    if (have_more_chunks || (_hwm != state._hwm)) {
```
Comment (tstuefe) [inline src/hotspot/share/memory/resourceArea.hpp:126]: If we started another chunk, _hwm cannot be state._hwm. So I would shorten this to `if  (_hwm != state._hwm)`. You can add an assert that the hwm should be different if we have follow up chunks.

Comment (shipilev) [reply src/hotspot/share/memory/resourceArea.hpp]: All right, done so.

File: src/hotspot/share/memory/resourceArea.hpp [hunk 2]
```diff
@@ -120,13 +122,28 @@ class ResourceArea: public Arena {
...
+
+      _chunk = state._chunk;
+      _hwm = state._hwm;
+      _max = state._max;
+
+      // Clear out this chunk (to detect allocation bugs).
+      // If we never got a new chunk, then we only need to clear
+      // up to replaced hwm.
+      if (ZapResourceArea) {
+        char* limit = have_more_chunks ? _max : replaced_hwm;
```
[the remark this answers is not in the document]
Comment (shipilev) [reply src/hotspot/share/memory/resourceArea.hpp]: Right. See new commit. This passes `tier1`, and I am re-running timing tests now...

**Review by tstuefe — APPROVED** [review]
Looks good!

**Comment by shipilev** [comment]
@coleenp, you would be interested in this as well, I think.

**Comment by coleenp** [comment]
Thanks @shipilev it does look very nice.  I'll read it carefully this afternoon.

**Review by coleenp — APPROVED** [review]
Ok, looks fine.  The comments help.  "replaced" didn't help me that much but can't think of a better name for it.

**Comment by shipilev** [comment]
> Ok, looks fine. The comments help. "replaced" didn't help me that much but can't think of a better name for it.

Had it like "old" originally, but it feels even more confusing: is it "old" that was before RM acquisition, or "old" that would be after RM release?

**Comment by coleenp** [comment]
yeah, keep "replaced" the comments tell you what it means.

**Comment by shipilev** [comment]
I ran `tier1..3` with Linux x86_64 release/fastdebug, and they pass. So, I am integrating.

/integrate
